File: netwalker_web/frontend/static/js/devices.js
// Device inventory page logic

let currentOffset = 0;
let currentFilters = {};
const PAGE_SIZE = 50;

// Load devices with current filters
async function loadDevices(offset = 0, append = false) {
    try {
        const params = new URLSearchParams({
            limit: PAGE_SIZE,
            offset: offset
        });
        
        // Add filters to params
        Object.keys(currentFilters).forEach(key => {
            if (currentFilters[key]) {
                params.append(key, currentFilters[key]);
            }
        });
        
        const response = await fetch(`/api/devices?${params.toString()}`);
        if (!response.ok) throw new Error('Failed to fetch devices');
        
        const devices = await response.json();
        const tbody = document.getElementById('devices-table');
        
        if (!append) {
            tbody.innerHTML = '';
        }
        
        if (devices.length === 0) {
            if (!append) {
                tbody.innerHTML = '<tr><td colspan="8" class="text-center">No devices found</td></tr>';
                document.getElementById('load-more').style.display = 'none';
            } else {
                document.getElementById('load-more').style.display = 'none';
            }
            updateShowingInfo(offset);
            return;
        }
        
        devices.forEach(device => {
            const row = document.createElement('tr');
            row.innerHTML = `
                <td>${device.device_name || '-'}</td>
                <td>${device.platform || '-'}</td>
                <td>${device.hardware_model || '-'}</td>
                <td>${device.ip_address || '-'}</td>
                <td>${device.serial_number || '-'}</td>
                <td>${device.software_version || '-'}</td>
                <td><small>${device.capabilities || '-'}</small></td>
                <td><span class="badge bg-${device.status === 'active' ? 'success' : 'secondary'}">${device.status}</span></td>
            `;
            tbody.appendChild(row);
        });
        
        // Show/hide load more button
        if (devices.length === PAGE_SIZE) {
            document.getElementById('load-more').style.display = 'block';
        } else {
            document.getElementById('load-more').style.display = 'none';
        }
        
        updateShowingInfo(offset + devices.length);
        
    } catch (error) {
        console.error('Error loading devices:', error);
        document.getElementById('devices-table').innerHTML = 
            '<tr><td colspan="8" class="text-center text-danger">Error loading devices</td></tr>';
    }
}

// Load device count
async function loadDeviceCount() {
    try {
        const params = new URLSearchParams();
        
        // Add filters to params
        Object.keys(currentFilters).forEach(key => {
            if (currentFilters[key]) {
                params.append(key, currentFilters[key]);
            }
        });
        
        const response = await fetch(`/api/devices/count?${params.toString()}`);
        if (!response.ok) throw new Error('Failed to fetch count');
        
        const data = await response.json();
        document.getElementById('total-count').textContent = data.count;
        
        // Show/hide filtered indicator
        if (data.filtered) {
            document.getElementById('filtered-indicator').style.display = 'inline-block';
        } else {
            document.getElementById('filtered-indicator').style.display = 'none';
        }
        
    } catch (error) {
        console.error('Error loading count:', error);
        document.getElementById('total-count').textContent = 'Error';
    }
}

// Update showing info
function updateShowingInfo(count) {
    document.getElementById('showing-info').textContent = `Showing ${count} devices`;
}

// Apply filters
function applyFilters(event) {
    event.preventDefault();
    
    // Get filter values
    currentFilters = {
        device_name: document.getElementById('filter-device-name').value.trim(),
        platform: document.getElementById('filter-platform').value.trim(),
        hardware_model: document.getElementById('filter-hardware-model').value.trim(),
        ip_address: document.getElementById('filter-ip-address').value.trim(),
        serial_number: document.getElementById('filter-serial-number').value.trim(),
        software_version: document.getElementById('filter-software-version').value.trim(),
        capabilities: document.getElementById('filter-capabilities').value.trim()
    };
    
    // Remove empty filters
    Object.keys(currentFilters).forEach(key => {
        if (!currentFilters[key]) {
            delete currentFilters[key];
        }
    });
    
    // Reset offset and reload
    currentOffset = 0;
    loadDevices(currentOffset);
    loadDeviceCount();
}

// Clear all filters
function clearFilters() {
    document.getElementById('filter-form').reset();
    currentFilters = {};
    currentOffset = 0;
    loadDevices(currentOffset);
    loadDeviceCount();
}

// Export to Excel
function exportToExcel() {
    const params = new URLSearchParams();
    
    // Add filters to params
    Object.keys(currentFilters).forEach(key => {
        if (currentFilters[key]) {
            params.append(key, currentFilters[key]);
        }
    });
    
    // Open export URL in new window
    const url = `/api/reports/devices?${params.toString()}`;
    window.open(url, '_blank');
}

// Initialize page
document.addEventListener('DOMContentLoaded', function() {
    // Load initial data
    loadDevices(currentOffset);
    loadDeviceCount();
    
    // Set up event listeners
    document.getElementById('filter-form').addEventListener('submit', applyFilters);
    document.getElementById('clear-filters').addEventListener('click', clearFilters);
    document.getElementById('export-button').addEventListener('click', exportToExcel);
    
    document.getElementById('load-more').addEventListener('click', function() {
        currentOffset += PAGE_SIZE;
        loadDevices(currentOffset, true);
    });
});

File: netwalker_web/backend/api/devices.py
"""
Device API endpoints

Author: Mark Oldham
"""

from fastapi import APIRouter, HTTPException, Query, Request
from typing import List, Optional
from pydantic import BaseModel

router = APIRouter()


class Device(BaseModel):
    """Device model"""
    device_id: int
    device_name: str
    platform: Optional[str] = None
    hardware_model: Optional[str] = None
    serial_number: Optional[str] = None
    capabilities: Optional[str] = None
    status: str
    software_version: Optional[str] = None
    ip_address: Optional[str] = None
    first_seen: Optional[str] = None
    last_seen: Optional[str] = None


@router.get("/devices", response_model=List[Device])
async def get_devices(
    request: Request,
    limit: int = Query(50, ge=1, le=1000),
    offset: int = Query(0, ge=0),
    device_name: Optional[str] = Query(None),
    platform: Optional[str] = Query(None),
    hardware_model: Optional[str] = Query(None),
    serial_number: Optional[str] = Query(None),
    capabilities: Optional[str] = Query(None),
    ip_address: Optional[str] = Query(None),
    software_version: Optional[str] = Query(None)
):
    """
    Get all devices with pagination and filtering
    
    Args:
        limit: Maximum number of devices to return (1-1000)
        offset: Number of devices to skip
        device_name: Filter by device name (partial match)
        platform: Filter by platform (partial match)
        hardware_model: Filter by hardware model (partial match)
        serial_number: Filter by serial number (partial match)
        capabilities: Filter by capabilities (partial match)
        ip_address: Filter by IP address (partial match)
        software_version: Filter by software version (partial match)
    
    Returns:
        List of devices matching filters
    """
    try:
        # Build filters dictionary
        filters = {}
        if device_name:
            filters['device_name'] = device_name
        if platform:
            filters['platform'] = platform
        if hardware_model:
            filters['hardware_model'] = hardware_model
        if serial_number:
            filters['serial_number'] = serial_number
        if capabilities:
            filters['capabilities'] = capabilities
        if ip_address:
            filters['ip_address'] = ip_address
        if software_version:
            filters['software_version'] = software_version
        
        devices = request.app.state.device_queries.get_all_devices(
            limit=limit, 
            offset=offset,
            filters=filters if filters else None
        )
        return devices
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/devices/{device_id}")
async def get_device(request: Request, device_id: int):
    """
    Get device details by ID
    
    Args:
        device_id: Device ID
    
    Returns:
        Device details
    """
    try:
        device = request.app.state.device_queries.get_device_by_id(device_id)
        if not device:
            raise HTTPException(status_code=404, detail="Device not found")
        return device
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/devices/search/{query}")
async def search_devices(
    request: Request,
    query: str,
    limit: int = Query(100, ge=1, le=1000)
):
    """
    Search devices by name, IP, or serial number
    
    Args:
        query: Search query
        limit: Maximum number of results
    
    Returns:
        List of matching devices
    """
    try:
        devices = request.app.state.device_queries.search_devices(query, limit=limit)
        return devices
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/devices/count")
async def get_device_count(
    request: Request,
    device_name: Optional[str] = Query(None),
    platform: Optional[str] = Query(None),
    hardware_model: Optional[str] = Query(None),
    serial_number: Optional[str] = Query(None),
    capabilities: Optional[str] = Query(None),
    ip_address: Optional[str] = Query(None),
    software_version: Optional[str] = Query(None)
):
    """
    Get total count of active devices with optional filters
    
    Returns:
        Device count
    """
    try:
        # Build filters dictionary
        filters = {}
        if device_name:
            filters['device_name'] = device_name
        if platform:
            filters['platform'] = platform
        if hardware_model:
            filters['hardware_model'] = hardware_model
        if serial_number:
            filters['serial_number'] = serial_number
        if capabilities:
            filters['capabilities'] = capabilities
        if ip_address:
            filters['ip_address'] = ip_address
        if software_version:
            filters['software_version'] = software_version
        
        count = request.app.state.device_queries.get_device_count(filters if filters else None)
        return {"count": count, "filtered": bool(filters)}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))
